### DISCUSSION TITLE
Feature: Reload local nodes

### DISCUSSION THREAD

**Discussion by Banbury** [IndiePython/nodezator#42] [Feature Request]
During the development of new nodes it would be convenient, if I could reload all local nodes at once.
All nodes in the current graph would need to be redrawn, and invalid connections deleted.
Auto reloading when a node is changed, would be even better.

**Comment by KennedyRichard** [comment]
Mr. Banbury,

Thank you for your feature request. All improvement suggestions are welcome and represent a chance of steering the app towards a successful path.

I'll properly reply to your suggestion in a day or two, this message is just so you know I'm aware of it.

I'll also take the liberty of converting your issue into a discussion, but don't worry about it, it doesn't mean your request is any less important. It is just that in Nodezator and other Indie Python repos we use issues for things that crash the app or otherwise prevent the user from doing something that is supposed to be available. When such an issue appears, any other work is paused so that it can be fixed ASAP.

This measure is taken for the benefit of the users: by doing things this way, whenever you have a real issue, it won't compete for space with other less urgent matters, just like your issue #40 that was given priority over other development tasks because it was a real issue (something that crashes the app).

A feature request is as important as an issue, but doesn't have the same urgency.

Peace.

**Comment by KennedyRichard** [comment]
Mr. Banbury,

I'm sorry for taking so long to answer you properly. Thank you for your patience.

On March 24th you suggested:

> During the development of new nodes it would be convenient, if I could reload all local nodes at once.
All nodes in the current graph would need to be redrawn, and invalid connections deleted.
Auto reloading when a node is changed, would be even better.

You can actually just save the file and hit F5. It will reload the file and the nodes as well.

However, since you mentioned the problem with invalid connections, then I assume you already know about that and also realized how invalid connections prevent Nodezator from opening a file.

The problem is that deleting the connections is not the only possibility, so Nodezator can't determine what to do with the connections by itself. The user might have deleted a parameter in a function, in which case the connection should indeed just be deleted, as you suggested, but the parameter might also just have been renamed, in which case we would not want the connection to be deleted, just point it to the renamed parameter. In other words, there is two valid possibilities, hence the conflict.

However, my answer to your suggestion is that it is a very good idea. A similar idea was also presented by a user on the discord server. Here's what he said:

> @Tom-99 [here](https://discord.com/channels/980193047000272967/980193047461630113/1013105326037409822):  I think a big help would be if Nodezator was able to open a ndz file that is outdated. Most nodes will still be perfectly fine. If Nodezator can mark the unstable nodes in red, for example, the ndz file can be repaired by the user. Now we have to edit the ndz file manually, which I am not yet able to do, or make the whole file from scratch (without the example of the existing file)

His comment led to a new feature being listed in our list of planned/requested features (#13). You can see it listed in the table as *creating a conflict resolution mode*. This was my response to his comment back then:

> me [here](https://discord.com/channels/980193047000272967/980193047461630113/1013183533587632298): I'm thinking of taking inspiration from how git handles conflicts, by marking the conflicts on the conflicted files, kinda like you already pointed. Nodezator could be improved to open a conflicted file and mark the conflicts for the user to solve and wouldn't allow the file to be edited until the conflicts are solved.

I still didn't have time to think about the specifics of the feature, but in principle, it would mark nodes with invalid connections so the user could quickly cycle through them, deciding whether the conflicted connection should be deleted (in case the parameter was really deleted) or be plugged into another parameter (in case of a renamed parameter).

The same would apply when nodes had their outputs renamed and/or the number of outputs changed.

However, you brought some new ideas to the table. That is, as you pointed out, sometimes people just want all the connections gone anyway. I'm thinking of allowing both options to be presented to the user when loading a file with conflicted connections in Nodezator: the user will be presented with a dialogue with two options:  1) delete all conflicting connections; and 2) open the file in conflict resolution mode.

I put a link to this discussion in the list of planned/requested features, besides the conflict resolution feature, so people can also see your suggestions and my reply here.

Of course, when it is time to implement this feature, I'll also open it for discussion as I always do, so other people can share their ideas as well. This will take some time since there are other more critical features to be implemented, but we'll tend to it as soon as reasonably possible. Until then, feel free to keep the ideas coming.

Again, thank you for your feedback/suggestions. We appreciate all feedback from our users: feature requests, suggestions, concerns, constructive criticism.

Peace.
